8472 Missing a Step in the CTF Calculation

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8472 
 - DESCRIBE PURPOSE OF THIS PULL REQUEST:
When looking through the code and comparing the state space method for calculating conduction transfer functions (CTFs) to the document that the code was based on (John Seem's doctoral dissertation, UW-Madison), a user noticed that a step was missing.  While it is unclear whether this was intentional or not, what is clear is that adding this step in will actually not make EnergyPlus more accurate.  Instead, it will actually have the potential to make it less accurate and potentially cause instabilities.  The missing step affects the calculation of the number of terms that must be calculated for an exponential matrix that is part of the CTF determination.  The missing step would actually reduce the number of terms and it isn't 100% clear that this missing step is necessary or that it is correct given the equations in this part of the code.  So, it was decided to leave the code as is (making this a documentation only bug fix--no new code, no unit tests) except for the addition of comments to clarify the difference between EnergyPlus and Seem's dissertation.  In addition, more documentation was added to the Engineering Reference to provide more detail about this portion of the calculation.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Tentative Fix

This simple commit includes a potential tentative fix to the issue.  This will be used as a test case and then further reorganization will happen if this is deemed successful.

RKStrand committed:
Correction after further code inspection

After looking at the code more thoroughly, made a correction to the correction.  This should be correct.

RKStrand committed:
Improved code comment

After some discussion with team members, it was decided that the code should be left as is and the comment in this section should be updated to reflect what is being done here.

RKStrand committed:
New documentation

New documentation for the issue.  Explains the calculation of the exponential matrix (briefly) and also justifies the decision to not expand

RKStrand committed:
Merge branch 'develop' into 8472MissingStepinCTFCalculation

mitchute commented on doc/engineering-reference/src/surface-heat-balance-manager-processes/conduction-through-the-walls.tex:
```diff
@@ -61,7 +61,27 @@ \subsection{Calculation of Conduction Transfer Functions}\label{calculation-of-c
+x_{t + \delta} = e^{A \delta} x_t + \int\limits_t^{t + \delta } {e^{A (t + \delta - \tau)} B u(\tau) d \tau}
+\end{equation}
+
+where t is a particular point in time, \(\delta}\) is the time step, and \(e^{A \delta}\) is the exponential matrix which is calculated using a power series:
```
Looks like there's a stray curly brace here ```\(\delta}\)```

RKStrand replied:
Yup--that's a stray.  Not sure how I missed that, but it doesn't surprise me that I accidentally had something like that in there.  Thanks for catching it and fixing it!

mitchute committed:
fix stray curly brace

mitchute reviewed (approved):
Thanks, @RKStrand! This should give the next interested reader (in another 20 years when E+ turns 40!) some additional details. There was one stray curly brace that I went ahead and fixed, but assuming docs build, this is good from my perspective.

mitchute commented:
@mjwitte do you want to take a look here before this goes in?

mjwitte commented on src/EnergyPlus/Construction.cc:
```diff
@@ -1197,7 +1197,25 @@ void ConstructionProps::calculateExponentialMatrix()
+    // Given that there is no harm in making l (integer of CheckVal) larger
+    // but it could result in a more accurate solution, it was decided to
+    // keep the value of AMatRowNormMax as it has been traditionally done
+    // in EnergyPlus since it could lead to higher accuracy.
```
My preference would be to just state the facts:



// exponentiation algorithm.  One thing to note is that the EnergyPlus    // code is slightly different here than what is presented in Seem's    // dissertation.  In Seem's dissertation, AMatRowNormMax (delta, the    // timestep, is already factored into this term above) is divided by    // 2^k with k defined in the code above.  ~This should have been done    // here to follow Seem's algorithm exactly.  However, in EnergyPlus,    // this was not done and there is some uncertainty as to why it wasn't.~    // Dividing AMatRowNormMax by 2^k would have the net effect of decreasing    // the value of AMatRowNormMax and thus potentially CheckVal as well.    // l, the integer value of CheckVal, is used to define the number of    // terms in the exponential matrix of AMat that are required for an    // accurate estimation of that matrix.  In practice, additional terms    // ~probably~ won't have a large effect but in general should make the    // calculation more accurate.  Also, if the new terms are too small,    // then as noted above the cut-off criteria will not use them.    ~// Given that there is no harm in making l (integer of CheckVal) larger    // but it could result in a more accurate solution, it was decided to    // keep the value of AMatRowNormMax as it has been traditionally done    // in EnergyPlus since it could lead to higher accuracy.~

RKStrand replied:
Sounds good, @mjwitte.  It is better to just stick with the facts.  I can save the editorializing for the incredibly exciting paper that I will publish on this topic (when I have nothing better to do 🤣).  I'll make this change (and the other) shortly.

mjwitte commented on doc/engineering-reference/src/surface-heat-balance-manager-processes/conduction-through-the-walls.tex:
```diff
@@ -61,7 +61,27 @@ \subsection{Calculation of Conduction Transfer Functions}\label{calculation-of-c
+L = minimum~of~[3 ||A \delta||_\infty + 6]~or~100
+\end{equation}
+
+where \(||A \delta||_\infty\) is the matrix row norm.  Seem, as part of his algorithm, indicates that the matrix row norm should be normalized so that it is less than unity by finding the power of 2 that is larger than the matrix row norm and dividing all elements of the A matrix by that term.  This is used later as part of the "scaling and squaring" approach to calculate the exponential matrix.  ...
```
A suggested edit:
~This seems somewhat unsual (especially given the equation for L has a limit of 100) and when Seem's approach was first implemented~ In EnergyPlus, the matrix row norm ~was~ *is* not scaled when ~it was~ used to calculate L.  ~So, while the actual implementation in EnergyPlus deviates slightly from what was recommended by Seem, this approach can perhaps be seen to reflect the improvements in computer technology since that work was published or simply that there appears to be a potential incongruency in the method that Seem suggested for that one detail.~ *This can result in additional terms which in general will make the calculation slightly more accurate.* ~In addition,~ To avoid excessive terms in the exponential matrix,  . . .

RKStrand replied:
I'll make these changes as requested also.  Thanks for reviewing all this, @mjwitte and @mitchute!

RKStrand committed:
Adressing reviewer comments

This commit addresses reviewer comments that were made about the new comments and documentation.  These changes should be exactly what the reviewer requested.

RKStrand committed:
Merge branch 'develop' into 8472MissingStepinCTFCalculation

mitchute commented:
Looks like GH Actions is having problems: https://www.githubstatus.com/incidents/m16jzl31gnqt

I built the Eng Ref locally, and it looks good. Merge at will @mjwitte.

mitchute commented:
GH Actions is back online and the docs built A-OK. Thanks, @RKStrand! Merging.
